src/core/qgsmaplayer.h

```diff
@@ -73,13 +73,17 @@ class CORE_EXPORT QgsMapLayer : public QObject
      */
     QgsMapLayer( QgsMapLayer::LayerType type = VectorLayer, const QString &name = QString::null, const QString &source = QString::null );
 
-    virtual ~QgsMapLayer();
+    /** Copy constructor.
+     * @param mapLayer the map layer to copy
+     * @param deep to copy unique id if true
+     */
+    QgsMapLayer( const QgsMapLayer &mapLayer, bool deep = true );
```

Comment:
@nyalldawson @m-kuhn 

I have some questions regarding the `clone` method implementation. 

As `QgsMapLayer` is a pure abstract class, I cannot have a `clone` method like:

```
QgsMapLayer* QgsMapLayer::clone()
{
  return new QgsMapLayer();
}
```

So I have to define a pure virtual method `clone` in `QgsMapLayer` and then implementing it in `QgsVectorLayer` and `QgsRasterLayer`. But doing this, I have to clone each attributes of `QgsMapLayer` within the clone of `QgsVectorLayer` and `QgsRasterLayer`. It's a lot of duplicated code (id, name, short name, and so on).

Or I could use a protected copy constructor but I'm not sure it's a good solution too...

I have the feeling that I'm missing something obvious...

Can you enlighten me?